sys/mount.c

```diff
@@ -64,30 +65,147 @@ int mount(const char *source, const char *target, const char *fstype, long mode,
 		return SET_ERRNO(-ENOTDIR);
 
 	msg.type = mtMount;
-	mnt->id = soid.id;
-	mnt->mode = mode;
-	strncpy(mnt->fstype, fstype, sizeof(mnt->fstype));
-	mnt->fstype[sizeof(mnt->fstype) - 1] = '\0';
+	imnt->dev = soid;
+	imnt->mnt = toid;
+	imnt->mode = mode;
+	strncpy(imnt->fstype, fstype, sizeof(imnt->fstype));
+	imnt->fstype[sizeof(imnt->fstype) - 1] = '\0';
 
 	msg.i.size = data != NULL ? strlen(data) : 0;
-	msg.i.data = data;
+	msg.i.data = (char *)data; /* FIXME: dropping const because of broken msg_t declaration */
 
 	if (msgSend(soid.port, &msg) < 0)
 		return SET_ERRNO(-EIO);
 
-	doidp = (oid_t *)msg.o.raw;
-	doid = *doidp;
+	/* Mount failed */
+	if (omnt->err < 0) {
+		return SET_ERRNO(omnt->err);
+	}
+	root = omnt->oid;
 
+	/* Create mountpoint */
 	memset(&msg, 0, sizeof(msg));
 
 	msg.type = mtSetAttr;
 	msg.i.attr.oid = toid;
 	msg.i.attr.type = atDev;
-	msg.i.data = &doid;
+	msg.i.data = &root;
 	msg.i.size = sizeof(oid_t);
 
 	if (msgSend(toid.port, &msg) < 0)
 		return SET_ERRNO(-EIO); /* FIXME: rollback partial mount */
 
 	return SET_ERRNO(msg.o.attr.err);
 }
+
+
+int umount(const char *target)
+{
+	msg_t msg = { 0 };
+	mount_o_msg_t *omnt = (mount_o_msg_t *)msg.o.raw;
+	int rootfs = 0;
+	oid_t oid, dev;
+	char *abspath;
+
+	abspath = resolve_path(target, NULL, 1, 0);
+	if (abspath == NULL) {
+		/* errno set by resolve_path() */
+		return -1;
+	}
+
+	if (lookup(abspath, &oid, &dev) < 0) {
+		free(abspath);
+		return SET_ERRNO(-ENOENT);
+	}
+
+	/* Check for attached device */
+	if (oid.port == dev.port) {
+		/* Check for rootfs */
+		if ((abspath[0] == '/') && (abspath[1] == '\0')) {
+			rootfs = 1;
+		}
+		/* Invalid target, neither device nor mountpoint */
+		else {
+			free(abspath);
+			return SET_ERRNO(-EINVAL);
+		}
+	}
+	free(abspath);
+
+	/* Check file type */
+	msg.type = mtGetAttr;
+	msg.i.attr.type = atMode;
+	msg.i.attr.oid = oid;
+
+	if (msgSend(oid.port, &msg) < 0) {
+		return SET_ERRNO(-EIO);
+	}
+
+	if (msg.o.attr.err < 0) {
+		return SET_ERRNO(msg.o.attr.err);
+	}
+
+	/* Got mountpoint */
+	if (S_ISDIR(msg.o.attr.val)) {
+		/* TODO: implement umount() by mountpoint, set ENOTSUP errno for now */
+		return SET_ERRNO(-ENOTSUP);
+	}
+	/* Got mounted device */
+	/* TODO: check device type (should be S_IFBLK?) */
+	else {
+		/* Get mountpoint */
+		msg.type = mtMountPoint;
+		msg.i.data = &dev;
+		msg.i.size = sizeof(dev);
+
+		if (msgSend(dev.port, &msg) < 0) {
+			return SET_ERRNO(-EIO);
+		}
+
+		if (omnt->err < 0) {
+			/* Check for rootfs */
+			if (omnt->err == -ENOENT) {
+				rootfs = 1;
```

Comment:
Hmm, why does this indicate rootfs? Perhaps it might, but it can also indicate other things I guess